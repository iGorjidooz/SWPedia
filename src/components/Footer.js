import React from 'react';
import { footer, footerContent, footerLink } from '../styles/FooterStyles';

const Footer = () => (
   <footer className={footer}>
      <div className={footerContent}>
         Created By <a className={footerLink} href="https://github.com/iGorjidooz">Iman Gorjidooz</a>
      </div>
   </footer>
);

export default Footer;