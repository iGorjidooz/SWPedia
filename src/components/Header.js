import React from 'react';
import { Link } from 'react-router-dom';
import * as generalStyles from '../styles/GeneralStyles';
import * as headerStyles from '../styles/HeaderStyle';

const Header = () => (
   <header className={headerStyles.header}>
      <div className={generalStyles.container}>
         <Link to="/" className={headerStyles.header__logo}>SWPedia</Link>
      </div>
   </header>
);

export default Header;