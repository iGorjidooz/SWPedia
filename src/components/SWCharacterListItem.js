import React from 'react';
import { Link } from 'react-router-dom';
import * as hStyles from '../styles/HelperStyles';
import * as cStyles from '../styles/ContentStyles';

const SWCharacterListItem = ({ character, listIndex }) => {
   return (
      <div className={cStyles.listItemWrapper}>
         <Link to={`/character/${character.id}`} className={cStyles.listItem}>
            <span className={cStyles.listIndex}>{listIndex}</span>
            <div>
               <h3 className={hStyles.margin0}>{character.name}</h3>
               <p className={hStyles.margin0}>Birth Year: {character.birth_year}</p>
               <p className={hStyles.margin0}>Gender: {character.gender}</p>
            </div>
         </Link>
      </div>
   );
};

export default SWCharacterListItem;