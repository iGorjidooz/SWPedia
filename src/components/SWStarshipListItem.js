import React from 'react';
import { Link } from 'react-router-dom';
import * as hStyles from '../styles/HelperStyles';
import * as cStyles from '../styles/ContentStyles';

const SWStarshipListItem = ({ starship, listIndex }) => {
   return (
      <div className={cStyles.listItemWrapper}>
         <Link to={`/starship/${starship.id}`} className={cStyles.listItem}>
            <span className={cStyles.listIndex}>{listIndex}</span>
            <div>
               <h3 className={hStyles.margin0}>{starship.name}</h3>
               <p className={hStyles.margin0}>Model: {starship.model}</p>
               <p className={hStyles.margin0}>Manufacturer: {starship.manufacturer}</p>
            </div>
         </Link>
      </div>
   );
};

export default SWStarshipListItem;