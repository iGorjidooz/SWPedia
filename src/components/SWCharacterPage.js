import React from 'react';
import Header from './Header';

const SWCharacterPage = () => (
   <div>
      <Header />
      <h1>SWCharacterPage</h1>
   </div>
);

export default SWCharacterPage;