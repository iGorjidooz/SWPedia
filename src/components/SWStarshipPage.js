import React from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { css } from 'emotion';
import Header from './Header';
import * as cStyles from '../styles/ContentStyles';
import * as gStyles from '../styles/GeneralStyles';

export const SWStarshipPage = (props) => {
   return props.starship === undefined ? (
      <Redirect to="/"/>
   ) : (
      <div className={gStyles.pageWrapper}>
         <Header />
         <div className={css`${gStyles.container};${cStyles.contentWrapper};${cStyles.infoWrapper}`}>
            <Link to="/" className={cStyles.backToListBtn}>&larr; Back to List</Link>
            <h1>{props.starship.name}</h1>
            <div className={cStyles.infoPanesWrapper}>
               <div className={cStyles.infoPane}>
                  <b>General Info:</b> <br/>
                  Model: {props.starship.model} <br/>
                  Manufacturer: {props.starship.manufacturer} <br/>
                  Starship Class: {props.starship.starship_class} <br/>
                  Cost in Credits: {props.starship.cost_in_credits} <br/>
                  Length: {props.starship.length}
               </div>
               <div className={cStyles.infoPane}>
                  <b>Performance:</b> <br/>
                  Max Atomsphering Speed: {props.starship.max_atmosphering_speed} <br />
                  Hyperdrive Rating: {props.starship.hyperdrive_rating} <br />
                  MGLT: {props.starship.MGLT} <br />
               </div>
               <div className={cStyles.infoPane}>
                  <b>Capacity:</b> <br />
                  Crew: {props.starship.crew} <br/>
                  Passengers: {props.starship.passengers} <br />
                  Cargo Capacity: {props.starship.cargo_capacity} <br />
                  Consumables: {props.starship.consumables}
               </div>
            </div>
         </div>
      </div>
   )
};

const mapStateToProps = (state, props) => ({
   starship: state.starships.items.find((starship) => starship.id === props.match.params.id)
});

export default connect(mapStateToProps)(SWStarshipPage);