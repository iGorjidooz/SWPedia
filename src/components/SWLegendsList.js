import React from 'react';
import { connect } from 'react-redux';
import { css } from 'emotion';
import SWStarshipListItem from './SWStarshipListItem';
import SWCharacterListItem from './SWCharacterListItem';
import { listItemWrapper, loadingSpinnerWrapper, loadingSpinner } from '../styles/ContentStyles';

export const SWLegendsList = (props) => {
   return (
      <div>
         {props.starshipItems.map((item, arrayIndex) => {
            const listIndex = arrayIndex + Math.floor(arrayIndex / 8) + 1;
            if (props.peopleItems.length > 0 && arrayIndex > 0 && (arrayIndex + 1) % 8 === 0) {
               const characterIndex = Math.floor(arrayIndex / 8);
               return (
                  <React.Fragment key={'listItemsWrapper' + arrayIndex}>
                     <SWStarshipListItem key={listIndex} starship={item} listIndex={listIndex} />
                     <SWCharacterListItem key={listIndex + 1} character={props.peopleItems[characterIndex]} listIndex={listIndex + 1} />
                  </React.Fragment>
               );
            } else {
               return (<SWStarshipListItem key={listIndex} starship={item} listIndex={listIndex} />);
            }
         })}
         {props.isFetching && <div className={css`${listItemWrapper};${loadingSpinnerWrapper}`}>
               <img className={loadingSpinner} src="/images/loader.gif" alt="Loading..." />
            </div>
         }
      </div>
   );
};

const mapStateToProps = (state) => ({
   starshipItems: state.starships.items,
   peopleItems: state.people.items,
   isFetching: state.starships.isFetching || state.people.isFetching
});

export default connect(mapStateToProps)(SWLegendsList);