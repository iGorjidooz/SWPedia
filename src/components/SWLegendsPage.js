import React from 'react';
import { connect } from 'react-redux';
import debounce from 'lodash/debounce';
import { css } from 'emotion';
import Header from './Header';
import SWLegendsList from './SWLegendsList';
import { startFetchingStarships } from '../actions/starshipsActions';
import { startFetchingPeople } from '../actions/peopleActions';
import * as gStyles from '../styles/GeneralStyles';
import * as cStyles from '../styles/ContentStyles';

export const SWLegendsPage = props => {
   // Scroll Event Handler Function
   // When the page scrolls to bottom it dispatches the actions to load next page of starships and people data, and unbind itself from event listener
   // throttle function from Lodash library has been used to improve performance of the scroll event handler
   const scrollEventHandler = (e) => {
      if(e.offsetHeight + e.scrollTop < (e.scrollHeight - 20)) {
         return;
      }
      if(!props.isFetching) {
         props.startFetchingStarships(props.nextStarshipsPageUrl);
         props.startFetchingPeople(props.nextPeoplePageUrl);
      }
   };

   const debouncedHandler = debounce(e => scrollEventHandler(e), 250, { maxWait: 500 });

   // Similar to componentDidUpdate lifecycle method
   // Looks for changes in props.isFetching and props.hasMoreStarships and binds or unbinds scrollEventHandler function to scroll event listener based on the condition
   // useLayoutEffect(() => {
   // //   props.hasMoreStarships && !props.isFetching
   // //     ? window.addEventListener("scroll", scrollEventHandler, {
   // //         passive: true
   // //       })
   // //     : window.removeEventListener("scroll", scrollEventHandler, {
   // //         passive: true
   // //       });
   // //   return () => {
   // //     window.removeEventListener("scroll", scrollEventHandler, {
   // //       passive: true
   // //     });
   // //   };
   // }, [props.isFetching, props.hasMoreStarships]);

   return (
     <div className={gStyles.pageWrapper}>
       <Header />
       <div className={css`${gStyles.container};${cStyles.contentWrapper}`}>
         <div
            id="legendsListWrapper"
            className={cStyles.listWrapper}
            onScroll={
            props.hasMoreStarships && !props.isFetching
               ? ({ currentTarget }) => debouncedHandler(currentTarget)
               : null
            }
         >
            <SWLegendsList />
         </div>
       </div>
     </div>
   );
};

const mapStateToProps = state => ({
   hasMoreStarships: !!state.starships.nextPageUrl,
   isFetching:  state.people.isFetching || state.starships.isFetching,
   nextStarshipsPageUrl: state.starships.nextPageUrl,
   nextPeoplePageUrl: state.people.nextPageUrl
});

const mapDispatchToProps = (dispatch) => ({
   startFetchingStarships: (nextStarshipsPageUrl) => dispatch(startFetchingStarships(nextStarshipsPageUrl)),
   startFetchingPeople: (nextPeoplePageUrl) => dispatch(startFetchingPeople(nextPeoplePageUrl))
 });

// Connecting the component to Redux store
export default connect(mapStateToProps, mapDispatchToProps)(SWLegendsPage);