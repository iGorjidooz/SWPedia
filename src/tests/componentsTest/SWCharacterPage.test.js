import React from 'react';
import { shallow } from 'enzyme';
import peopleFixtures from '../fixtures/peopleFixtures';
import { SWCharacterPage } from '../../components/SWCharacterPage';

test('should redirect to homepage if the id passed to the page is not found in the store.people array', () => {
  const wrapper = shallow(
    <SWCharacterPage character={undefined} listIndex={peopleFixtures[1].id} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render SWCharacterPage correctly', () => {
  const wrapper = shallow(
    <SWCharacterPage character={peopleFixtures[1]} listIndex={peopleFixtures[1].id} />
  );
  expect(wrapper).toMatchSnapshot();
});