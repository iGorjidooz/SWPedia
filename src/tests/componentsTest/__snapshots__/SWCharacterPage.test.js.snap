// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should redirect to homepage if the id passed to the page is not found in the store.people array 1`] = `
<Redirect
  to="/"
/>
`;

exports[`should render SWCharacterPage correctly 1`] = `
<div
  className="css-eswrbj"
>
  <Header />
  <div
    className="css-jfh2bv"
  >
    <Link
      className="css-rd4qh1"
      to="/"
    >
      ‚Üê Back to List
    </Link>
    <h1>
      C-3PO
    </h1>
    <div
      className="css-1806dpm"
    >
      <div
        className="css-1wetx9n"
      >
        <b>
          Profile:
        </b>
         
        <br />
        Height: 
        167
         
        <br />
        Mass: 
        75
         
        <br />
        Hair Color: 
        n/a
         
        <br />
        Skin color: 
        gold
         
        <br />
        Eye Color: 
        yellow
         
        <br />
        Birth Year: 
        yellow
         
        <br />
        Gender: 
        n/a
         
        <br />
      </div>
    </div>
  </div>
</div>
`;
